ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/akkastreams/dispatcher/SignalDispatcher.scala

```diff
@@ -63,27 +63,46 @@ class SignalDispatcher private () {
   /** Closes this SignalDispatcher by gracefully completing the existing Source subscriptions.
     * For any downstream with pending signals, at least one such signal will be sent first.
     */
-  def shutdown(): Future[Unit] = shutdownInternal(_.complete())
+  def shutdown(): Future[Unit] =
+    shutdownInternal { source =>
+      source.complete()
+      source.watchCompletion()
+    }
 
   /** Closes this SignalDispatcher by failing the existing Source subscriptions with the provided throwable. */
-  def fail(throwable: Throwable): Future[Unit] =
-    shutdownInternal(_.fail(throwable)).transform {
-      // This throwable is expected so map to Success
-      case Failure(`throwable`) => Success(())
-      case other => other
-    }(ExecutionContext.parasitic)
+  def fail(throwableBuilder: () => Throwable): Future[Unit] =
+    shutdownInternal { source =>
+      implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.parasitic
+      val throwable = throwableBuilder()
+      source.fail(throwable)
+      source
+        .watchCompletion()
+        .recoverWith {
+          // This throwable is expected so map to Success
+          case `throwable` => Future.unit
+        }
+    }
 
   private def shutdownInternal(
-      shutdownSourceQueue: SourceQueueWithComplete[_] => Unit
+      shutdownSourceQueue: SourceQueueWithComplete[_] => Future[_]
   ): Future[Unit] = {
     implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.parasitic
     runningState
       .getAndSet(None)
       .fold(throw new IllegalStateException("SignalDispatcher is already closed")) { sources =>
```

Comment:
I know it is not this PR but why is this not a Future.failed instead?

Reply:
will change.